### PR TITLE
[8.x] Implemented the trackable events for the control dispatching process

### PR DESCRIPTION
**Pull request by smskin**
Added functionality that allows you to track the lifecycle of events (in my project it is used for logging the event bus).

### CODE REVIEW

**Comment by smskin** [comment]
@kafkiansky please tell me why you disliked this feature? Do you know a better solution to the event bus event tracking problem?

File: src/Illuminate/Events/Dispatcher.php [hunk 1]
```diff
@@ -265,9 +272,48 @@ public function dispatch($event, $payload = [], $halt = false)
...
+                $payload,
+                $halt
+            )
+        );
+    }
+
+    private function trackEventDispatched($event, array $payload, bool $halt)
+    {
+        if (!$this->isTrackableEvent($event))
+        {
```
Comment (mateusjunges) [inline src/Illuminate/Events/Dispatcher.php:299]:
```suggestion
        if (!$this->isTrackableEvent($event)) {
```

File: src/Illuminate/Events/Dispatcher.php [hunk 2]
```diff
@@ -265,9 +272,48 @@ public function dispatch($event, $payload = [], $halt = false)
...
         }
 
+        $this->trackEventDispatched($event, $payload, $halt);
+
         return $halt ? null : $responses;
     }
 
+    private function trackEventDispatching($event, array $payload, bool $halt)
+    {
+        if (!$this->isTrackableEvent($event))
```
Comment (mateusjunges) [inline src/Illuminate/Events/Dispatcher.php:282]:
```suggestion
        if (! $this->isTrackableEvent($event)) {
```

File: src/Illuminate/Events/Dispatcher.php [hunk 3]
```diff
@@ -265,9 +272,48 @@ public function dispatch($event, $payload = [], $halt = false)
...
+            return;
+        }
+
+        $this->dispatch(
+            new EventDispatching(
+                $event,
+                $payload,
+                $halt
+            )
+        );
```
Comment (mateusjunges) [inline src/Illuminate/Events/Dispatcher.php:293]:
```suggestion
        $this->dispatch(new EventDispatching($event, $payload, $halt));
```

File: src/Illuminate/Events/Dispatcher.php (hunk 1 above)
Comment (smskin) [reply src/Illuminate/Events/Dispatcher.php]: fixed

File: src/Illuminate/Events/Dispatcher.php (hunk 2 above)
Comment (smskin) [reply src/Illuminate/Events/Dispatcher.php]: fixed

File: src/Illuminate/Events/Dispatcher.php (hunk 3 above)
Comment (smskin) [reply src/Illuminate/Events/Dispatcher.php]: fixed

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

**Pull request closed without merging**